src/shooting.jl

```diff
@@ -0,0 +1,142 @@
+`retract(M, p, X, retraction) ≈ q`.
+
+This implementation of the shooting method works by using another inverse retraction to form
+the first guess of the vector. This guess is updated by shooting the vector, guessing the
+vector pointing from the shooting result to the target point, and transporting this vector
+update back to the initial point on aa discretized grid. This process is repeated until the
+norm of the vector update falls below a specified tolerance or the maximum number of
+iterations is reached.
+
+# Keywords
+- `initial_inverse_retraction::AbstractInverseRetractionMethod`: The inverse retraction used
+    to form the initial guess of the vector. Defaults to
+    [`default_inverse_retraction`](@ref)`(M)`
+- `vector_transport::AbstractVectorTransportMethod`: The vector transport used to transport
+    the initial guess of the vector. Defaults to [`default_vector_transport`](@ref)`(M)`.
+- `num_transport_points::Int=2`: The number of discretization points used for vector
+    transport in the shooting method. 2 is the minimum number of points, including just the
+    endpoints.
+- `tolerance::Real`: The tolerance for the shooting method. Defaults to `sqrt(eps())` for
+    the eltypes of `p` and `q`.
+- `max_iterations::Int=10_000`: The maximum number of iterations for the shooting method.
+"""
+function shooting(
+    M::AbstractManifold,
+    p,
+    q,
+    retraction::AbstractRetractionMethod;
+    kwargs...,
+)
+    X = allocate_result(M, inverse_retract, p, q)
+    shooting!(M, X, p, q, retraction; kwargs...)
+    return X
+end
+function shooting!(
+    M::AbstractManifold,
+    X,
+    p,
+    q,
+    retraction::AbstractRetractionMethod;
+    initial_inverse_retraction=default_inverse_retraction_method(M),
+    vector_transport=default_vector_transport_method(M),
+    num_transport_points::Int=2,
+    tolerance::Real=sqrt(eps(float(real(eltype(X))))),
+    max_iterations::Int=10_000,
+)
+    inverse_retract!(M, X, p, q, initial_inverse_retraction)
+    gap = norm(M, p, X)
+    gap < tolerance && return X
+    T = real(Base.promote_eltype(X, p, q))
+    transport_grid = range(one(T), zero(T); length=num_transport_points)[2:(end - 1)]
+    ΔX = allocate(X)
+    ΔXnew = tX = allocate(ΔX)
+    retr_tX = allocate_result(M, retract, p, X)
+    if num_transport_points > 2
+        retr_tX_new = allocate_result(M, retract, p, X)
+    end
+    iteration = 1
+    while (gap > tolerance) && (iteration < max_iterations)
+        retract!(M, retr_tX, p, X, retraction)
+        inverse_retract!(M, ΔX, retr_tX, q, initial_inverse_retraction)
+        gap = norm(M, retr_tX, ΔX)
+        for t in transport_grid
+            tX .= t .* X
+            retract!(M, retr_tX_new, p, tX, retraction)
+            vector_transport_to!(M, ΔXnew, retr_tX, ΔX, retr_tX_new, vector_transport)
+            # realias storage
+            retr_tX, retr_tX_new, ΔX, ΔXnew, tX = retr_tX_new, retr_tX, ΔXnew, ΔX, ΔX
+        end
```

Comment:
This seems like it should be its own generic vector transport method. That is, given `(p, X, q)` and a `retraction_method` such that `q = retr(M, p, X)`, then the method extends a local vector transport to be longer range by incrementally transporting along the discretization given by `t -> retr(M, p, t*X)`.

Reply:
I think we could to this with something like the following:
```julia

struct VectorTransportAlongRetraction
    M
    p
    X
    ts
    retraction
    vector_transport
end

struct VectorTransportAlong
    M
    c
    vector_transport
end

function vector_transport_to!(
    M::AbstractManifold,
    Y,
    p,
    X,
    q,
    method::VectorTransportAlongRetraction,
)
    c = map(t -> retract(M, p, t * X, method.retraction))
    return vector_transport_to!(M, Y, p, X, q, VectorTransportAlong(M, c, method.vector_transport))
end

function vector_transport_to!(
    M::AbstractManifold,
    Y,
    p,
    X,
    q,
    method::VectorTransportAlong,
)
    c = [method.c..., q]
    return vector_transport_along!(M, Y, p, X, c, method.vector_transport)
end
```

Then we would declare the vector transport used here with
```julia
VectorTransportAlongRetraction(
    M,
    p,
    X,
    range(1, 0; length=num_transport_points)[2:end],
    retraction,
    vector_transport,
)
```
and this loop is just replace by a simple `vector_transport_to!` call.

Reply:
We have these incremental ideas already in pole ladder and schild‘s ladder – and yes that is more like a vector transport to ;)

Reply:
To be clear, I want to just have `vector_transport_to!` in `shooting!`, and I'm trying to come up with missing generic `struct`s that would allow me to perform this specific vector transport using just a single `vector_transport_to!` call.

Can you give an example of how pole ladder and schild's ladder would accomplish this with the current `struct`s? I could not figure out a way.

Reply:
ah then I misread the last part of your previous message which I read as: the whole stuff can (instead!) be done by a simple vector transport to, but then I am a little confused that your first function is a `_to` function and not an `along` one.

Both pole and schild are little different approaches to your step wise vector transport, but they are also `_to` methods that discretise the path along the way (given e.g. by an retraction) and then for every “step” perform something (which is not a vector transport but an approximation of that). So what you do can not be done with them since they approximate the single step, but they discretise the same way you do.

Reply:
The challenge is that when we call `shooting`, we don't know what `X` is, and we don't know where the initial "shot" (result of the retraction for the initial `X`) will be. So we have no way of specifying the discretization. So we can either define the discretization in the method as we have done here, or we can define new vector transport types that only implicitly define a discretization.

And I just realized my above types won't work for this anyways, since they require knowing `X` ahead of time, which we won't. So perhaps defining the discretization as done here is the best way, since setting `num_transport_points=2` ignores the discretization anyways.

Reply:
yes I think just specifying the discretisation would be the best way to go then